lib/auth/auth_with_roles.go

```diff
@@ -1326,18 +1326,24 @@ func (a *ServerWithRoles) checkUnifiedAccess(resource types.ResourceWithLabels,
 		return false, trace.Wrap(canAccessErr)
 	}
 
-	if resourceKind != types.KindSAMLIdPServiceProvider {
-		if err := checker.CanAccess(resource); err != nil {
+	// Filter first and only check RBAC if there is a match to improve perf.
+	match, err := services.MatchResourceByFilters(resource, filter, nil)
+	if !match || err != nil {
+		return false, trace.Wrap(err)
+	}
```

Comment:
Now that we are filtering before evaluating RBAC, we don't want to open some mechanism by which someone could probe for information about resources by constructing malicious filters than only generate errors in certain conditions (afaik thats not currently possible, but we have to pretend that it is I think).

In theory we could construct a system where we continue on with access checks and only yield the error if the access checks pass, but I think a simpler mechanism would be to just treat match errors the way we do deserialization errors: generate a warn log and skip the item, but never bubble the error up to the caller.